Xamarin.Forms.Core/Routing.cs

```diff
@@ -8,25 +9,73 @@ public static class Routing
 	{
 		static int s_routeCount = 0;
 		static Dictionary<string, RouteFactory> s_routes = new Dictionary<string, RouteFactory>();
+		static Dictionary<string, Page> s_implicitPageRoutes = new Dictionary<string, Page>();
 
 		const string ImplicitPrefix = "IMPL_";
 		const string DefaultPrefix = "D_FAULT_";
-		const string _pathSeparator = "/";
+		internal const string PathSeparator = "/";
+
+		// We only need these while a navigation is happening 
+		internal static void ClearImplicitPageRoutes()
+		{
+			s_implicitPageRoutes.Clear();
+		}
+
+		internal static void RegisterImplicitPageRoute(Page page)
+		{
+			var route = GetRoute(page);
+			if (!IsUserDefined(route))
+				s_implicitPageRoutes[route] = page;
+		}
+
+		// Shell works much better if the entire nav stack can be represented by a string
+		// If the users pushes pages without using routes we want these page keys tracked
+		internal static void RegisterImplicitPageRoutes(Shell shell)
+		{
+			foreach(var item in shell.Items)
+			{
+				foreach(var section in item.Items)
+				{
+					for(int i = 1; i < section.Navigation.NavigationStack.Count; i++)
+					{
```

Comment:
With this change, we will have a performance improvement. I did a test and against mono, this can be 12% faster
```suggestion
				{
				var navigationStackCount = section.Navigation.NavigationStack.Count;
					for(int i = 1; i < navigationStackCount ; i++)
					{
```